prober/http.go

```diff
@@ -0,0 +1,62 @@
+// Copyright (c) 2022 Tailscale Inc & AUTHORS All rights reserved.
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+package prober
+
+import (
+	"bytes"
+	"context"
+	"fmt"
+	"io"
+	"net/http"
+)
+
+const maxHTTPBody = 4 << 20 // MiB
+
+// HTTP returns a Probe that healthchecks an HTTP URL.
+//
+// The Probe sends a GET request for url, expects an HTTP 200
+// response, and verifies that want is present in the response
+// body. If the URL is HTTPS, the probe further checks that the TLS
+// certificate is good for at least the next 7 days.
+func HTTP(url, wantText string) Probe {
+	return func(ctx context.Context) error {
+		return probeHTTP(ctx, url, []byte(wantText))
+	}
+}
+
+func probeHTTP(ctx context.Context, url string, want []byte) error {
+	req, err := http.NewRequestWithContext(ctx, "GET", url, nil)
+	if err != nil {
+		return fmt.Errorf("constructing request: %w", err)
+	}
+
+	// Get a completely new transport each time, so we don't reuse a
+	// past connection.
+	tr := http.DefaultTransport.(*http.Transport).Clone()
+	defer tr.CloseIdleConnections()
+	c := &http.Client{
+		Transport: tr,
+	}
+
+	resp, err := c.Do(req)
+	if err != nil {
+		return fmt.Errorf("fetching %q: %w", url, err)
```

Comment:
maybe known error types so that we can have varz?

Reply:
Hmm. This is a bit terrible to export in a form Prometheus will understand, so to send richer alerts we'd have to speak the alertmanager protocol directly, and make the prober fully stateful. I'd prefer to avoid that for now, but if we don't have that, this prober needs an HTTP UI where you can ask it about recent history. I'll think about how to do this.

Reply:
I just meant like export counters for the number of times we've seen a type of known error. Nothing too fancy

Reply:
Hmm, maybe, but being a periodic prober, it's kinda weird for it to keep track of errors seen over its lifetime, unconnected from the probe cycles? Maybe? I'm not sure.

I think I'd like to refactor the probing API a little bit if we want this, to let probes set their own vars. Okay if I do in a followup?

Reply:
yeah, can add later as needed as well. This is fine for now!